src/controller/CHIPDeviceController.cpp

```diff
@@ -1175,6 +1184,47 @@ void DeviceCommissioner::OnFailedToExtendedArmFailSafeDeviceAttestation(void * c
     commissioner->CommissioningStageComplete(CHIP_ERROR_INTERNAL, report);
 }
 
+void DeviceCommissioner::OnICDSymmetricKeyGenerationCompleted(void * context, CHIP_ERROR error, NodeId controllerNodeId,
+                                                              uint64_t subjectId, ICDRegistrationDelegate::ICDKey key,
+                                                              Optional<ICDRegistrationDelegate::ICDKey> verificationKey)
+{
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(Controller, "Failed to generate symmetric key for ICD registration. Error %s", ErrorStr(error));
+        commissioner->CommissioningStageComplete(error);
+        return;
+    }
+    if (commissioner->mDeviceBeingCommissioned == nullptr)
+    {
+        ChipLogError(Controller, "No device is being commissioned at this time.");
+        commissioner->CommissioningStageComplete(CHIP_ERROR_INTERNAL);
+        return;
+    }
+    CommissioningDelegate::CommissioningReport report;
+    ICDSymmetricKeyInfo info = ICDSymmetricKeyInfo{
+        .controllerNodeId = controllerNodeId,
+        .subjectId        = subjectId,
+        .key              = key,
+        .verificationKey  = verificationKey,
+    };
+    report.Set<ICDSymmetricKeyInfo>(info);
+    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
+}
+
+void DeviceCommissioner::OnICDManagementRegisterClientResponse(
+    void * context, const app::Clusters::IcdManagement::Commands::RegisterClientResponse::DecodableType & data)
+{
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+
+    CommissioningDelegate::CommissioningReport report;
+    ICDRegistrationInfo info;
+    info.icdCounter = data.ICDCounter;
+    report.Set<ICDRegistrationInfo>(info);
+    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
```

Comment:
where do you plan to save ICDSymmetricKeyInfo and starting counter into persistent storage? inside CommissioningStageComplete?

Reply:
We should do it in OnRegistrationComplete.